### PR TITLE
Propagate create statistics

### PR DESCRIPTION
**Pull request by agedemenli**
DESCRIPTION: `CREATE STATISTICS` statements will be propagated across workers.

fixes: #4391

### CODE REVIEW

**Review by marcocitus — COMMENTED** [review]
High-level note: We also need to create already existing statistics when creating new shards (e.g. in create_distributed_table, or when moving a shard).

File: src/backend/distributed/commands/statistics.c [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+	Oid relationId = RangeVarGetRelid(relation, AccessExclusiveLock, false);
+
+	if (!IsCitusTable(relationId))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	QualifyTreeNode((Node *) stmt);
```
Comment (halilozanakgul) [inline src/backend/distributed/commands/statistics.c:65]: We don't use stmt later, why do we qualify it?

File: src/backend/distributed/relay/relay_event_utility.c [hunk 2]
```diff
@@ -522,6 +522,28 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
...
+			SetSchemaNameIfNotExist(objectSchemaName, schemaName);
+			AppendShardIdToName(relationName, shardId);
+
+			Value *statNameValue = llast(createStatsStmt->defnames);
+			char *statName = strVal(statNameValue);
+			char **statNameVar = &(statName);
+			AppendShardIdToName(statNameVar, shardId);
+			statNameValue = makeString(pstrdup(statName));
+			createStatsStmt->defnames = list_make2(makeString(pstrdup(schemaName)),
+												   statNameValue);
```
Comment (halilozanakgul) [inline src/backend/distributed/relay/relay_event_utility.c:542]: Using the schemaName would cause statistics to always be in the same schema with the table but they don't have to be.

You can also add some related tests.

File: src/test/regress/sql/propagate_statistics.sql [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    b   int
+);
+
+CREATE STATISTICS s2 (dependencies) ON a, b FROM test_stats;
+
+SELECT create_distributed_table('test_stats2', 'a');
+
+DROP TABLE test_stats;
+DROP TABLE test_stats2;
+DROP SCHEMA "statistics'test";
```
Comment (halilozanakgul) [inline src/test/regress/sql/propagate_statistics.sql:32]: If you add CASCADE here you can remove DROP TABLE statements.

File: src/backend/distributed/commands/statistics.c [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+	systable_endscan(scanDescriptor);
+	table_close(pgStatistics, NoLock);
+
+	return statisticsIdList;
+}
+
+
+static char *
+pg_get_statisticsobj_worker(Oid statextid, bool missing_ok)
```
Comment (halilozanakgul) [inline src/backend/distributed/commands/statistics.c:160]: We might want to add this function to ruleutils instead

**Review by halilozanakgul — CHANGES_REQUESTED** [review]

**Comment by halilozanakgul** [comment]
Maybe not in this PR but we will need to propagate the `DROP STATISTICS` queries too.

File: src/backend/distributed/relay/relay_event_utility.c (hunk 2 above)
Comment (agedemenli) [reply src/backend/distributed/relay/relay_event_utility.c:542]: Good point. Fixed. Thanks

File: src/backend/distributed/commands/statistics.c (hunk 4 above)
Comment (agedemenli) [reply src/backend/distributed/commands/statistics.c:160]: I thought it would be appropriate to keep it among other functions we use for `CREATE STATISTICS`, all in `statistics.s`. In the future if we need to use this function for some other reason (which I don't expect), we can move it to a more common place like `ruleutils`.

What do you think @halilozanakgul ? Any idea for some another case we need this func?

Comment (halilozanakgul) [reply src/backend/distributed/commands/statistics.c:160]: All of the `pg_get_*` functions are in the ruleutils files. I asked to move this too to keep consistent.

Comment (agedemenli) [reply src/backend/distributed/commands/statistics.c:160]: hmm that makes sense. alright then

File: src/test/regress/sql/propagate_statistics.sql [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+-- test for distributing an already existing statistics
+CREATE TABLE test_stats2 (
+    a   int,
+    b   int
+);
+
+CREATE STATISTICS s2 (dependencies) ON a, b FROM test_stats;
+
+SELECT create_distributed_table('test_stats2', 'a');
+
```
Comment (halilozanakgul) [inline src/test/regress/sql/propagate_statistics.sql:29]: Can you add some tests with schemas. For example:

- Table and statistics in different schemas
- Current schema isn't public

File: src/test/regress/sql/propagate_statistics.sql [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+-- test create statistics propagation
+CREATE TABLE test_stats (
+    a   int,
+    b   int
+);
+
+SELECT create_distributed_table('test_stats', 'a');
+
+-- SET citus.log_remote_commands TO true;
```
Comment (halilozanakgul) [inline src/test/regress/sql/propagate_statistics.sql:17]: If this will remain commented out we can remove this.

File: src/test/regress/expected/propagate_statistics.out [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+CREATE TABLE test_stats2 (
+    a   int,
+    b   int
+);
+CREATE STATISTICS s2 (dependencies) ON a, b FROM test_stats;
+SELECT create_distributed_table('test_stats2', 'a');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
```
Comment (halilozanakgul) [inline src/test/regress/expected/propagate_statistics.out:30]: There are no checks to see if the statistics are propagated correctly. Not all bugs cause an error message.

**Review by halilozanakgul — CHANGES_REQUESTED** [review]

**Comment by halilozanakgul** [comment]
We have to also propagate the schema of the statistics too.

```
CREATE SCHEMA sc1;
CREATE TABLE tbl (a int, b int);
SELECT create_distributed_table ('tbl', 'a');

CREATE STATISTICS sc1.st1 ON a, b FROM tbl;

ERROR:  schema "sc1" does not exist
CONTEXT:  while executing command on localhost:9702
```

**Comment by marcocitus** [comment]
> Maybe not in this PR but we will need to propagate the DROP STATISTICS queries too.

I think DROP STATISTICS should be in scope, otherwise it becomes very awkward to test.

File: src/backend/distributed/commands/statistics.c [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+ * PostprocessCreateStatisticsStmt is called after a CREATE STATISTICS command has
+ * been executed by standard process utility.
+ */
+List *
+PostprocessCreateStatisticsStmt(Node *node, const char *queryString)
+{
+	bool missingOk = false;
+	ObjectAddress objectAddress = GetObjectAddressFromParseTree(node, missingOk);
+
+	EnsureDependenciesExistOnAllNodes(&objectAddress);
```
Comment (marcocitus) [inline src/backend/distributed/commands/statistics.c:94]: this should be unnecessary, statistics cannot (currently) depend on other objects

also, beware that we would go through this code path on the workers as well, which would cause deadlocks.

File: src/backend/distributed/commands/statistics.c [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+ * CREATE STATISTICS.
+ */
+List *
+PreprocessCreateStatisticsStmt(Node *node, const char *queryString)
+{
+	CreateStatsStmt *stmt = castNode(CreateStatsStmt, node);
+
+	DDLJob *ddlJob = palloc0(sizeof(DDLJob));
+	RangeVar *relation = (RangeVar *) linitial(stmt->relations);
+	Oid relationId = RangeVarGetRelid(relation, AccessExclusiveLock, false);
```
Comment (marcocitus) [inline src/backend/distributed/commands/statistics.c:57]: we should make sure to use the right lock type here to avoid blocking reads

CreateStatistics in statscmds.c only uses ShareUpdateExclusiveLock

File: src/backend/distributed/commands/statistics.c (hunk 8 above)
Comment (marcocitus) [reply src/backend/distributed/commands/statistics.c:103]: discussed over chat, statistics can actually depend on schemas which may be different from the table schema

File: src/backend/distributed/commands/statistics.c [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+ * been executed by standard process utility.
+ */
+List *
+PostprocessCreateStatisticsStmt(Node *node, const char *queryString)
+{
+	CreateStatsStmt *stmt = castNode(CreateStatsStmt, node);
+	Assert(stmt->type == T_CreateStatsStmt);
+
+	if (!IsCoordinator() || !ShouldPropagate())
+	{
```
Comment (marcocitus) [inline src/backend/distributed/commands/statistics.c:96]: This does not cover the case where the shard is on the coordinator.

File: src/backend/distributed/commands/statistics.c [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+/*
+ * PreprocessCreateStatisticsStmt is called during the planning phase for
+ * CREATE STATISTICS.
+ */
+List *
+PreprocessCreateStatisticsStmt(Node *node, const char *queryString)
+{
+	CreateStatsStmt *stmt = castNode(CreateStatsStmt, node);
+
+	DDLJob *ddlJob = palloc0(sizeof(DDLJob));
```
Comment (marcocitus) [inline src/backend/distributed/commands/statistics.c:55]: minor: probably makes sense to allocate this later (e.g. just before `ddlJob->targetRelationId = `)

it looks a little cleaner and we don't do unnecessary memory allocation (e.g. if this is not a Citus table).

File: src/backend/distributed/commands/statistics.c (hunk 10 above)
Comment (marcocitus) [reply src/backend/distributed/commands/statistics.c:96]: We could maybe repeat the IsCitusTable check, or set the EnableDDLPropagation GUC in worker_apply_ddl_command.

File: src/backend/distributed/commands/statistics.c (hunk 9 above)
Comment (agedemenli) [reply src/backend/distributed/commands/statistics.c]: thank you! fixed

File: src/backend/distributed/commands/statistics.c [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+	EnsureCoordinator();
+
+	QualifyTreeNode((Node *) stmt);
+
+	char *ddlCommand = DeparseTreeNode((Node *) stmt);
+
+	DDLJob *ddlJob = palloc0(sizeof(DDLJob));
+
+	ddlJob->targetRelationId = RangeVarGetRelid(relation, ShareUpdateExclusiveLock,
```
Comment (halilozanakgul) [inline src/backend/distributed/commands/statistics.c:70]: Cannot you use relationId here?

File: src/backend/distributed/commands/statistics.c [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	List *createStatisticsCommandList = NIL;
+
+	PushOverrideEmptySearchPath(CurrentMemoryContext);
+
+	List *statisticsIdList = GetExplicitStatisticsIdList(relationId);
+
+	Oid statisticsId = InvalidOid;
+	foreach_oid(statisticsId, statisticsIdList)
+	{
+		char *createStatisticsCommand = pg_get_statisticsobj_worker(statisticsId, true);
```
Comment (halilozanakgul) [inline src/backend/distributed/commands/statistics.c:148]: If missing_ok is true we might need to check if the result is NULL or not. (or maybe missing_ok needs to be false)

File: src/backend/distributed/deparser/qualify_statistics_stmt.c [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+		if (searchPath == NIL)
+		{
+			appendStringInfo(&schemaName, "public");
+		}
+		else
+		{
+			Oid schemaId = linitial_oid(searchPath);
+			char *fetchedNamespaceName = get_namespace_name(schemaId);
+			appendStringInfo(&schemaName, "%s", fetchedNamespaceName);
```
Comment (halilozanakgul) [inline src/backend/distributed/deparser/qualify_statistics_stmt.c:56]: You can use `appendStringInfoString` here and for the one before.
But also why not just use a char*, we do not append more than once?

File: src/backend/distributed/relay/relay_event_utility.c [hunk 15]
```diff
@@ -522,6 +522,41 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
...
+
+			if (containsSchemaName)
+			{
+				createStatsStmt->defnames = list_make2(linitial(
+														   createStatsStmt->defnames),
+													   statNameValue);
+			}
+			else
+			{
+				createStatsStmt->defnames = list_make1(statNameValue);
```
Comment (halilozanakgul) [inline src/backend/distributed/relay/relay_event_utility.c:554]: Can the code ever come here? If so why did we not need to add the schema here too?

File: src/backend/distributed/relay/relay_event_utility.c [hunk 16]
```diff
@@ -522,6 +522,41 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
...
+
+			Value *statNameValue = llast(createStatsStmt->defnames);
+			char *statName = strVal(statNameValue);
+			char **statNameVar = &(statName);
+			AppendShardIdToName(statNameVar, shardId);
+			statNameValue = makeString(pstrdup(statName));
+
+			bool containsSchemaName = list_length(createStatsStmt->defnames) > 1;
+
+			if (containsSchemaName)
```
Comment (halilozanakgul) [inline src/backend/distributed/relay/relay_event_utility.c:546]: This looks a bit weird to me. Can you put the comparison here in the if statement and a very small comment above , like `check if it contains schema name`

File: src/test/regress/sql/propagate_statistics.sql [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+CREATE STATISTICS s1 (dependencies) ON a, b FROM test_stats;
+
+-- test for distributing an already existing statistics
+CREATE TABLE test_stats2 (
+    a   int,
+    b   int
+);
+
+CREATE STATISTICS s2 (dependencies) ON a, b FROM test_stats;
```
Comment (halilozanakgul) [inline src/test/regress/sql/propagate_statistics.sql:25]: This should be test_stats2.

File: src/test/regress/sql/propagate_statistics.sql [hunk 18]
```diff
@@ -0,0 +1,42 @@
+CREATE SCHEMA "statistics'test";
+
+SET search_path TO "statistics'test";
+SET citus.next_shard_id TO 980000;
+SET client_min_messages TO WARNING;
+SET citus.shard_count TO 32;
```
Comment (halilozanakgul) [inline src/test/regress/sql/propagate_statistics.sql:6]: If you want you can have less shard_count. So the output will be smaller.

**Review by halilozanakgul — CHANGES_REQUESTED** [review]

File: src/backend/distributed/commands/statistics.c (hunk 12 above)
Comment (agedemenli) [reply src/backend/distributed/commands/statistics.c:70]: yes, thanks

File: src/backend/distributed/commands/statistics.c (hunk 13 above)
Comment (agedemenli) [reply src/backend/distributed/commands/statistics.c:148]: I think it should be false

File: src/backend/distributed/relay/relay_event_utility.c (hunk 15 above)
Comment (agedemenli) [reply src/backend/distributed/relay/relay_event_utility.c:554]: No, it seems the code will never reach here, added this part just to be defensive. Didn't add a schema name because we don't have one, in this case

File: src/backend/distributed/relay/relay_event_utility.c (hunk 16 above)
Comment (agedemenli) [reply src/backend/distributed/relay/relay_event_utility.c:546]: sure

File: src/test/regress/sql/propagate_statistics.sql (hunk 17 above)
Comment (agedemenli) [reply src/test/regress/sql/propagate_statistics.sql:25]: missed, thanks

File: src/backend/distributed/deparser/qualify_statistics_stmt.c [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+		}
+		else
+		{
+			Oid schemaId = linitial_oid(searchPath);
+			schemaName = get_namespace_name(schemaId);
+		}
+
+		Value *statName = linitial(stmt->defnames);
+		stmt->defnames = list_make2(makeString(pstrdup(schemaName)), statName);
+	}
```
Comment (thanodnl) [inline src/backend/distributed/deparser/qualify_statistics_stmt.c:59]: Here you have implemented a lot of complex behaviour of Postgres choosing a namespace. This is non-trivial and more complicated than what you have implemented.

- the first namespace on the path might not have creation permissions for the current user
- it is unknown if the `public` schema is the actual fallback if no `search_path` is set.

Just two things on the top of my mind.

Instead, I would treat the `stmt->defnames` as a `NameList`. This is a datastructure often used in postgres when identifying objects. You can convert that to a `RangeVar`.

`RangeVarGetCreationNamespace` can tell you the actual namespace/schema postgres will create objects in. Set this on the `RangeVar` and convert the `RangeVar` back to a `NameList`

The conversion functions you need are:
 - `makeRangeVarFromNameList`
 - `MakeNameListFromRangeVar`

File: src/backend/distributed/deparser/ruleutils_11.c [hunk 20]
```diff
@@ -7488,6 +7489,122 @@ pg_get_triggerdef_command(Oid triggerId)
 	return pg_get_triggerdef_worker(triggerId, prettyOutput);
 }
 
+char *
+pg_get_statisticsobj_worker(Oid statextid, bool missing_ok)
```
Comment (thanodnl) [inline src/backend/distributed/deparser/ruleutils_11.c:7493]: I feel this function comes from postgres directly right?
Will not review it

File: src/backend/distributed/deparser/ruleutils_12.c [hunk 21]
```diff
@@ -7487,6 +7488,126 @@ pg_get_triggerdef_command(Oid triggerId)
 	bool prettyOutput = false;
 	return pg_get_triggerdef_worker(triggerId, prettyOutput);
 }
+char *
```
Comment (thanodnl) [inline src/backend/distributed/deparser/ruleutils_12.c:7491]: We need 2 empty lines above here.

Our formatter ignores these files as they come mostly from postgres. However, lets keep some simple rules applied here.

File: src/backend/distributed/deparser/ruleutils_11.c [hunk 22]
```diff
@@ -7488,6 +7489,122 @@ pg_get_triggerdef_command(Oid triggerId)
 	return pg_get_triggerdef_worker(triggerId, prettyOutput);
 }
 
+char *
```
Comment (thanodnl) [inline src/backend/distributed/deparser/ruleutils_11.c:7492]: can we add 1 more line above here. See comment on next function

File: src/backend/distributed/relay/relay_event_utility.c [hunk 23]
```diff
@@ -522,6 +522,40 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
...
+			char **objectSchemaName = &(relation->schemaname);
+
+			SetSchemaNameIfNotExist(objectSchemaName, schemaName);
+			AppendShardIdToName(relationName, shardId);
+
+			Value *statNameValue = llast(createStatsStmt->defnames);
+			char *statName = strVal(statNameValue);
+			char **statNameVar = &(statName);
+			AppendShardIdToName(statNameVar, shardId);
+			statNameValue = makeString(pstrdup(statName));
```
Comment (thanodnl) [inline src/backend/distributed/relay/relay_event_utility.c:542]: this seems very complex code for what we want to do.
currently it finds the original pointer to where we store the string pointer so to inplace append the shardid.
then we copy the value out of it to create our new `NameList`.

Can we instead copy the `strVal` somewhere local and pass the address of that into `AppendShardIdToName`. This will not destroy or alter the original memory while also making it easier to follow.

```suggestion
			Value *statNameValue = llast(createStatsStmt->defnames);
			char *statName = strVal(statNameValue);
			statName = pstrdup(statName)
			AppendShardIdToName(&statName, shardId);
```

File: src/backend/distributed/relay/relay_event_utility.c (hunk 15 above)
Comment (thanodnl) [reply src/backend/distributed/relay/relay_event_utility.c:553]: For this whole block and the above, I would again do the transformations between `NameList` -> `RangeVar` -> `append shard id name in range var` -> convert back to `NameList`.

It also solves the unreachable parts of the code. And shows more the intention of the code.

**Review by thanodnl — CHANGES_REQUESTED** [review]

File: src/backend/distributed/deparser/ruleutils_11.c (hunk 20 above)
Comment (agedemenli) [reply src/backend/distributed/deparser/ruleutils_11.c:7493]: yes, copy-pasted

**Comment by thanodnl** [comment]
Schema's housing a statistic for a table are not created before distributing a table:

```sql
[local] nilsdijk@postgres:9700-36841=# CREATE TABLE foo (a int, b text);
CREATE TABLE
Time: 4.105 ms

[local] nilsdijk@postgres:9700-36841=# CREATE SCHEMA bar;
CREATE SCHEMA
Time: 6.202 ms

[local] nilsdijk@postgres:9700-36841=# CREATE STATISTICS bar.baz ON a, b FROM foo;
CREATE STATISTICS
Time: 3.266 ms

[local] nilsdijk@postgres:9700-36841=# SELECT create_distributed_table('foo', 'a');
ERROR:  schema "bar" does not exist
CONTEXT:  while executing command on localhost:9702
Time: 63.850 ms
```

File: src/backend/distributed/deparser/deparse_statistics_stmts.c [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+AppendColumnNames(StringInfo buf, CreateStatsStmt *stmt)
+{
+	ListCell *cell = NULL;
+	foreach(cell, stmt->exprs)
+	{
+		Node *node = (Node *) lfirst(cell);
+		Assert(IsA(node, ColumnRef));
+
+		ColumnRef *column = (ColumnRef *) node;
+		Assert(list_length(column->fields) == 1);
```
Comment (thanodnl) [inline src/backend/distributed/deparser/deparse_statistics_stmts.c:116]: I was able to hit this Assert with a 'simple' query:

```sql
CREATE STATISTICS bar ON foo.a FROM foo;
```

This is unsupported syntax from postgres' perspective, however as you can see this quite easily could cause edgecases in our deparser to hit.

We can do 2 things in these cases.

 - Return the same error as postgres would return
 - actually deparse into the same ddl statement I typed.

My suggestion would be to do the later.

In this case it is a simple loop over `column->fields` and print them as a qualified identifiers, separated with dots.

As expected, postgres already has us covered with `NameListToQuotedString`.

File: src/backend/distributed/deparser/deparse_statistics_stmts.c [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+
+		char *columnName = strVal((Value *) linitial(column->fields));
+
+		appendStringInfoString(buf, columnName);
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/deparser/deparse_statistics_stmts.c:120]:
```sql
[local] nilsdijk@postgres:9700-61144=# CREATE TABLE foo(a int, "Bar" text);
CREATE TABLE
Time: 47.247 ms

[local] nilsdijk@postgres:9700-61144=# SELECT create_distributed_table('foo', 'a');
┌──────────────────────────┐
│ create_distributed_table │
├──────────────────────────┤
│                          │
└──────────────────────────┘
(1 row)

Time: 51.515 ms

[local] nilsdijk@postgres:9700-61144=# CREATE STATISTICS bar ON a, "Bar" FROM foo;
ERROR:  column "bar" does not exist
CONTEXT:  while executing command on localhost:9701
Time: 12.204 ms
```

As it seems we need quoting here!
This will be solved when using `NameListToQuotedString` here.

File: src/test/regress/expected/propagate_statistics.out [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+     2
+(1 row)
+
+\c - - - :master_port
+DROP SCHEMA "statistics'test" CASCADE;
+NOTICE:  drop cascades to 3 other objects
+DETAIL:  drop cascades to table "statistics'test".test_stats
+drop cascades to table "statistics'test".test_stats2
+drop cascades to table "statistics'test".tbl
+DROP SCHEMA sc1 CASCADE;
```
Comment (thanodnl) [inline src/test/regress/expected/propagate_statistics.out:107]: nit; To prevent this output changing everything you add anything to the schema you could call `SET client_min_messages TO warning;` right before the drop. That will hide these messages.

File: src/backend/distributed/commands/type.c [hunk 27]
```diff
@@ -258,7 +258,7 @@ PreprocessCreateEnumStmt(Node *node, const char *queryString)
 	QualifyTreeNode(node);
 
 	/* reconstruct creation statement in a portable fashion */
-	const char *createEnumStmtSql = DeparseCreateEnumStmt(node);
+	const char *createEnumStmtSql = DeparseTreeNode((Node *) node);
```
Comment (thanodnl) [inline src/backend/distributed/commands/type.c:261]: unclear why this change is made here.

That said, why are we casting to a `Node *` if it is already that type?

File: src/backend/distributed/deparser/qualify_statistics_stmt.c [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+
+	RangeVar *stat = makeRangeVarFromNameList(stmt->defnames);
+
+	if (stat->schemaname == NULL)
+	{
+		Oid schemaOid = RangeVarGetCreationNamespace(stat);
+		stat->schemaname = get_namespace_name(schemaOid);
+
+		stmt->defnames = MakeNameListFromRangeVar(stat);
+	}
```
Comment (thanodnl) [inline src/backend/distributed/deparser/qualify_statistics_stmt.c:49]: Love how small and understandable this became compared with earlier versions!

File: src/test/regress/sql/propagate_statistics.sql [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+CREATE STATISTICS s2 (dependencies) ON a, b FROM test_stats2;
+
+SELECT create_distributed_table('test_stats2', 'a');
+
+-- test when stats is on a different schema
+CREATE SCHEMA sc1;
+CREATE TABLE tbl (a int, b int);
+SELECT create_distributed_table ('tbl', 'a');
+
+CREATE STATISTICS sc1.st1 ON a, b FROM tbl;
```
Comment (thanodnl) [inline src/test/regress/sql/propagate_statistics.sql:34]: Can we add a test where the statistics on a _new_ schema are created before we distribute the table. As this showed a problem during my last review it is good to have it verified that regression does not slip in anymore.

File: src/test/regress/sql/propagate_statistics.sql [hunk 30]
```diff
@@ -0,0 +1,43 @@
+CREATE SCHEMA "statistics'test";
```
Comment (thanodnl) [inline src/test/regress/sql/propagate_statistics.sql:1]: I miss some tests that would fail if quoting mistakes were made.

This like tables with columns having capital letters, which require proper quotes.

This also needs to be performed twice, once with the statistics created before the table gets distributed and once after.

Same would be for a schemaname that requires quotes, and tablenames that would require quotes.

The different quoted things (schema, columns, tables) can all be in the same test. However we should exercise them on both codepaths for the sql ddl generation.

File: src/test/regress/sql/propagate_statistics.sql [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+
+-- test when stats is on a different schema
+CREATE SCHEMA sc1;
+CREATE TABLE tbl (a int, b int);
+SELECT create_distributed_table ('tbl', 'a');
+
+CREATE STATISTICS sc1.st1 ON a, b FROM tbl;
+
+\c - - - :worker_1_port
+SELECT stxname FROM pg_statistic_ext ORDER BY stxname ASC;
```
Comment (thanodnl) [inline src/test/regress/sql/propagate_statistics.sql:37]: this test easily fails if any other test add statistics and runs before this test. Could we scope this to only show the statistics for the schema's we use in this test suite?

```sql
SELECT stxname
FROM pg_statistic_ext
WHERE stxnamespace IN (
	SELECT oid
	FROM pg_namespace
	WHERE nspname IN ('public')
);
```

with the ordering ofc, and the appropriate namespaces.

File: src/backend/distributed/commands/statistics.c [hunk 32]
```diff
@@ -0,0 +1,209 @@
...
+	PushOverrideEmptySearchPath(CurrentMemoryContext);
+
+	List *statisticsIdList = GetExplicitStatisticsIdList(relationId);
+
+	Oid statisticsId = InvalidOid;
+	foreach_oid(statisticsId, statisticsIdList)
+	{
+		ObjectAddress address = { 0 };
+		ObjectAddressSet(address, StatisticExtRelationId, statisticsId);
+		EnsureDependenciesExistOnAllNodes(&address);
```
Comment (thanodnl) [inline src/backend/distributed/commands/statistics.c:149]: As discussed via chat.

Instead of creating distributed transactions here, please add the schema's of the statistics as an expansion on the relation they are on:

https://github.com/citusdata/citus/blob/d3019f1b6d153177cc9d64fa4b8190a655e7bc57/src/backend/distributed/metadata/dependency.c#L914-L929

File: src/test/regress/sql/propagate_statistics.sql [hunk 33]
```diff
@@ -17,27 +17,51 @@ SELECT create_distributed_table('test_stats', 'a');
...
 
-CREATE STATISTICS sc1.st1 ON a, b FROM tbl;
+CREATE STATISTICS sc1.st1 ON a, "B" FROM tbl;
+
+-- test distributing table with already created stats on a new schema
+CREATE TABLE test_stats3 (
+    a int,
+    b int
+);
+CREATE SCHEMA sc2;
```
Comment (thanodnl) [inline src/test/regress/sql/propagate_statistics.sql:41]: :+1: great to use a new one here! otherwise it could have been propagated above

**Review by thanodnl — APPROVED** [review]
LGTM
